e2e_tests/tests/experiment/test_custom_searcher.py

```diff
@@ -567,47 +772,96 @@ def promote_async(self, request_id: uuid.UUID, metric: float) -> List[Operation]
+                if (
+                    not trial_metric.promoted
+                    and trial_metric.request_id not in self.asha_search_state.closed_trials
+                ):
+                    if trial_metric.request_id not in self.asha_search_state.early_exit_trials:
                         logging.info(f"Closing trial {trial_metric.request_id}")
                         ops.append(Close(trial_metric.request_id))
-                        self.closed_trials.add(trial_metric.request_id)
+                        self.asha_search_state.closed_trials.add(trial_metric.request_id)
         return ops
 
     def sample_params(self) -> Dict[str, object]:
         hparams = {
             "global_batch_size": 10,
-            "metrics_base": 0.05 * (len(self.trial_rungs) + 1),
+            "metrics_base": 0.05 * (len(self.asha_search_state.trial_rungs) + 1),
             "metrics_progression": "constant",
         }
         logging.info(f"hparams={hparams}")
         return hparams
 
     def progress(self) -> float:
-        if 0 < self.max_concurrent_trials < self.pending_trials:
+        if 0 < self.asha_search_state.max_concurrent_trials < self.asha_search_state.pending_trials:
             raise RuntimeError("Pending trial is greater than max concurrent trials")
+        all_trials = len(self.asha_search_state.rungs[0].metrics)
 
-        all_trials = len(self.rungs[0].metrics)
-        progress = all_trials / (1.2 * self.max_trials)
-        if all_trials == self.max_trials:
-            num_valid_trials = self.completed_trials - self.invalid_trials
-            progress_no_overhead = num_valid_trials / self.max_trials
+        progress = all_trials / (1.2 * self.asha_search_state.max_trials)
+        if all_trials == self.asha_search_state.max_trials:
+            num_valid_trials = (
+                self.asha_search_state.completed_trials - self.asha_search_state.invalid_trials
+            )
+            progress_no_overhead = num_valid_trials / self.asha_search_state.max_trials
             progress = max(progress_no_overhead, progress)
 
         return progress
+
+    def save_method_state(self, path: Path) -> None:
+        self.raise_exception("save_method_state")
+        checkpoint_path = path.joinpath("method_state")
+        with checkpoint_path.open("wb") as f:
+            pickle.dump(self.asha_search_state, f)
+
+    def load_method_state(self, path: Path) -> None:
+        self.raise_exception("load_method_state")
+        checkpoint_path = path.joinpath("method_state")
+        with checkpoint_path.open("rb") as f:
+            self.asha_search_state = pickle.load(f)
+
+    def raise_exception(self, exception_id: str) -> None:
+        if exception_id == self.exception_point:
+            logging.info(f"Raising exception in {exception_id}")
+            ex = MaxRetryError(HTTPConnectionPool(host="dummyhost", port=8080), "http://dummyurl")
+            raise ex
+
+
+class MockLocalSearchRunner(LocalSearchRunner):
```

Comment:
"Mock" suggests to _me_ that you might be using a mocking library, like `unittest` or `mock`. Again, not feeling very strongly about it but maybe call it "FailableLocalSearchRunner" or "FallibleSearchRunner" :)